auth/client/src/main/java/org/wildfly/security/auth/client/DynamicSSLContextImpl.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2020 Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.wildfly.security.auth.client;
+
+import org.kohsuke.MetaInfServices;
+import org.wildfly.security.dynamic.ssl.DynamicSSLContextException;
+import org.wildfly.security.dynamic.ssl.DynamicSSLContextSPI;
+
+import javax.net.ssl.SSLContext;
+import java.net.URI;
+import java.security.AccessController;
+import java.security.GeneralSecurityException;
+import java.security.PrivilegedAction;
+import java.util.List;
+
+// implementation of dynamic ssl context by elytron client
+@MetaInfServices(value = DynamicSSLContextSPI.class)
+public class DynamicSSLContextImpl implements DynamicSSLContextSPI {
+
+    private final AuthenticationContextConfigurationClient AUTH_CONTEXT_CLIENT =
+            AccessController.doPrivileged((PrivilegedAction<AuthenticationContextConfigurationClient>) AuthenticationContextConfigurationClient::new);
+    private AuthenticationContext authenticationContext = AuthenticationContext.captureCurrent();
+    private SSLContext configuredDefaultSSLContext;
+    private List<SSLContext> configuredSSLContexts;
+
+    public DynamicSSLContextImpl() throws GeneralSecurityException {
+        this.configuredSSLContexts = AUTH_CONTEXT_CLIENT.getConfiguredSSLContexts(authenticationContext);
```

Comment:
To make use of a dynamic SSL context for an outbound connection on the server side, it looks like `authentication-client` configuration would need to be specified with appropriate SSL contexts and `match-host` and `match-port` rules. One thing that was mentioned in the description for EAP7-1121 was possibly being able to specify a dynamic SSL context for the JVM wide default SSL context. What would this configuration look like on the server side?

Reply:
@fjuma Thanks a lot for review. Yes, we certainly want to use this to be able to configure a JVM wide default SSL context and it should be possible to use it on the server side for outbound connections. 

> it looks like authentication-client configuration would need to be specified with appropriate SSL contexts and `match-host` and `match-port` rules.

Where do you think `authentication-client` configuration should be specified?
I might be wrong and missing something, but it seemed to me that since I am using current AUTH_CONTEXT_CLIENT for obtaining context  - eg. `AUTH_CONTEXT_CLIENT.getSSLContext(uri, authenticationContext)` and `AuthenticationContext.captureCurrent()` - that means that whatever `authentication-client` is active that one will be used, and so I don't need to specify `authentication-client` configuration in the code. So it will be dynamic thanks to captureCurrent when it is being used. But I think I should move `AuthenticationContext.captureCurrent()` inside the call: `AUTH_CONTEXT_CLIENT.getSSLContext(uri, authenticationContext)`. 

If this won't work like I expect here, maybe on server side there can be different implementation of `DynamicSSLContextSPI` that will implement methods `SSLContext getConfiguredDefault();`, `List<SSLContext> getConfiguredSSLContexts();`,  `SSLContext getSSLContext(URI uri) throws DynamicSSLContextException;` differently. But I think usage of these methods can stay the same. Am I answering your question here?

Reply:
Since `DynamicSSLContextImpl` makes use of `AUTH_CONTEXT_CLIENT.getConfiguredSSLContexts`, to configure a dynamic SSL context for an outbound connection on the server side, it looked like `authentication-client` configuration would need to be specified in the Elytron subsystem configuration. I think maybe it would be good to have an analysis document to look at along with these changes (it doesn't need to be submitted against the wildfly-proposals repo yet) just to get a better sense of how a dynamic SSL context would be used on the server side and client side.

In order to configure a JVM wide default SSL context on the server side, we set the `default-ssl-context` attribute to a previously configured `ssl-context`. The main question I have is what would the configuration for that SSL context look like on the server side in order to be able to specify a dynamic SSL context there?